### ISSUE TITLE
Wrapper around the telegram.keyboard object (keyboard_markup, reply and inline)

### ISSUE THREAD

**Issue by david-shiko** [python-telegram-bot/python-telegram-bot#4608]
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

I want to use the.keyboard as a set of buttons rather than a nested list of rows, which are hard to manage and modify.
Indeed, I don't care which row or column contains my button.
Rows and columns are the technically unique identifiers of the button, but for human readability, the button's actual text is better because there can't be two identical buttons in the same keyboard.
In my opinion, it would be a nice fit into the `telegram.ext` package as the next generation of the keyboard object that we already have.
Prerequisites:

Looking for a proper button object in a nested loop is always tedious and requires some concentration and time.
Most medium and large PTB bots require a custom Button class to respect the DRY principle and keep the code clean.
Let's do it out of the box!

### Describe the solution you'd like

I think it would be nice if you would add this feature.
See the possible usage example:

Notice:
`keyboard = update.message.keyboard  # incoming keyboard`
Return type depend on the PTB policy because PTB objects are immutable.

```
# Text and callback modification example
keyboard.buttons.order_the_pizza.edit_text(new_text='pizza succussfuly ordered')
# or
keyboard.buttons.order_the_pizza.text = 'pizza succussfuly ordered'

# Callback modification example:
keyboard.buttons.some_emoji.callback = 'foo'
# Another possible method names: change order | update_index | set_priority, etc.
```
```
# Pattern example:
# P.S. Read only access to the patterns
print(keyboard.pattern)
print(keyboard.buttons.coffee.pattern)
```
```
# Order modification example:
keyboard.buttons.breaking_news.change_position(new_row_index=-1, new_column_index=2)
# Another possible method names: change order | update_index, etc.

# Another sample:
last_button_text = keyboard.buttons[-1][-1].text  # index access are also possible.
```

```
# Order aliases:
# Even more progressive way: buttons.last means buttons[-1][-1]
# The "last" name reserved but does the "last" may mean something else except the "last" button itself?
last_button = keyboard.buttons.last

# The same with a row:
last_row = keyboard.rows.last
first_column = keyboard.columns.first
fifty_column = keyboard.columns.fifty
```

```
from dataclasses import dataclass
from telegram import InlineKeyboardButton


@dataclass  # The struct of the button
class ButtonWrapper(InlineKeyboardButton):
    row_index: int  # new
    column_index: int  # new
    text: int
    callback: int
    ...  # all other regular attrs of InlineKeyboardButton

# Many possible useful methods and shortcuts 


```
### Internal implementation: ###
1. The wrapper itself resembles the `SimpleNamespace` Python class or the `Dataclass`.
2. Order aliases (first, second, fifth, etc.) resemble the `Enum`.
3. Order aliases may accept an optional strict policy as a parameter.
Soft - If the count of buttons is less than the order alias (only two buttons in a row, for example), return the second button.
Mediate - returns `None`.
Hard - raises `IndexError`.

**Comment by Bibo-Joshi** [comment]
Hi. Thanks for reaching out.

However I'm having a hard time trying to understand what you're getting at 😅
Overall I have the impression that in your use cases you *receive* a keyboard and then want to

* inspect what buttons are in there
* modify the keyboard

Is that about right? If so, can you please clarify what exactly the individual code blocks above are for? Can you give example handler callbacks that show how those would be used?

My first general thoughts on your descriptions are:

* keep in mind that not every button press gives you access to the keyboard that the button belongs to. think of keyboards sent via inline mode.
* instead of inspecting the received keyboard and making decisions based on that, I highly recommend to use the `callback_data` as sole base for any decisions
* instead of modifying a keyboard in-place I would also use an approach of this kind:
    1. modify a dataset based on the received `callback_data`
    2. build a new keyboard based on the modified data set
    3. send that new keyboard

---
> because there can't be two identical buttons in the same keyboard.

why not?

**Comment by david-shiko** [comment]
> Overall I have the impression that in your use cases you receive a keyboard and then want to

- inspect what buttons are in there
- modify the keyboard

Exactly!

> keep in mind that not every button press gives you access to the keyboard that the button belongs to. think of keyboards sent via inline mode.

Okay, I will keep in mind this.

> instead of inspecting the received keyboard and making decisions based on that, I highly recommend to use the callback_data as sole base for any decisions

sometimes you actually no need callback data for example if your buttons are checkboxes.

> instead of modifying a keyboard in-place I would also use an approach of this kind

Under the hood this will be indeed a new keyboard object.

> because there can't be two identical buttons in the same keyboard.

> why not?

How you will distinct them if text the same in 2 buttons, by the row/column order? Anyway You can set alias as 'first_duplicated_btn', "second_duplicated", etc

P.S. I see that you are not impressed with such wrapper, maybe this will good fit into contrlib?

**Comment by Bibo-Joshi** [comment]
> > Overall I have the impression that in your use cases you receive a keyboard and then want to
> 
> * inspect what buttons are in there
> * modify the keyboard
> 
> Exactly!

Nice :) I'd still be interested in example handler callbacks to better understand your desired interfaces.

> > instead of modifying a keyboard in-place I would also use an approach of this kind
> 
> Under the hood this will be indeed a new keyboard object.

This was not obvious to me from your example snippets. Handler callback examples might help to understand this better.

> > > because there can't be two identical buttons in the same keyboard.
> 
> > why not?
> 
> How you will distinct them if text the same in 2 buttons, by the row/column order? Anyway You can set alias as 'first_duplicated_btn', "second_duplicated", etc

My point is: The TG API has no issue with two buttons having the same text within one keyboard. There might be valid use cases for that. If you still want to differentiate between them, can just use  set uniqe `callback_data` values rather than relying on row/column parsing :)

> P.S. I see that you are not impressed with such wrapper, maybe this will good fit into contrlib?

At least so far I don't see a generic/broad use case so far 😅 But as I said above, I don't have the feeling that I understood the interface yet that you're looking for.
A contribution to ptbcontrib is always an option, sure.

**Comment by Bibo-Joshi** [comment]
@david-shiko you haven't replied with more detailed examples yet. If You want to continue the discussion, please provide the requested examples. Otherwise I'll have to close.

**Comment by david-shiko** [comment]
> [@david-shiko](https://github.com/david-shiko) you haven't replied with more detailed examples yet. If You want to continue the discussion, please provide the requested examples. Otherwise I'll have to close.

Yes, you can close it. I hope to make a PR, but not in the near future.
